src/IECoreScene/ShaderNetworkAlgo.cpp

```diff
@@ -684,6 +713,213 @@ IECore::DataPtr loadSpline(
+
+					adapterArraySize = splineValues.size();
+
+					if( adapterArraySize <= maxArrayInputAdapterSize )
+					{
+						for( unsigned int i = 0; i < splineValues.size(); i++ )
+						{
+							adapter->parameters()[ g_arrayInputNames[i] ] = new Color3fData( splineValues[i] );
+						}
+					}
+				}
+				else
+				{
+					knotOffset = getEndPointDuplication( floatSplineData->readable().basis ).first;
+					adapter = new Shader( "Utility/FloatArrayInputAdapter", "osl:shader" );
+
+					const FloatVectorData *splineValuesData = findValues != newParameters.end() ? runTimeCast<const FloatVectorData>( findValues->second.get() ) : nullptr;
+					if( !splineValuesData )
+					{
+						throw IECore::Exception( "Internal failure in convertToOSLConventions - expandSpline did not create values" );
+					}
+
+					const std::vector< float > &splineValues = splineValuesData->readable();
+
+					adapterArraySize = splineValues.size();
+
+					if( adapterArraySize <= maxArrayInputAdapterSize )
+					{
+						for( unsigned int i = 0; i < splineValues.size(); i++ )
+						{
+							adapter->parameters()[ g_arrayInputNames[i] ] = new FloatData( splineValues[i] );
+						}
+					}
+				}
+
+				if( adapterArraySize > maxArrayInputAdapterSize )
+				{
+					throw IECore::Exception(
+						"Cannot handle input to " + connection.destination.shader.string() + "." +
+						destName + ". Expanded spline has " + std::to_string( adapterArraySize) +
+						" control points, but max input adapter size is " + std::to_string( maxArrayInputAdapterSize )
+					);
+				}
+
+				// TODO - does this need to be tagged to be reversible?
+				// ANSWER - it's a special node type, which users shouldn't be creating, so John pointed out
+				// we can probably just remove them all without needing to tag
+				//adapter->blindData()->writable()[ componentConnectionAdapterLabel() ] = g_trueData;
+
+				adapterHandle = network->addShader( s.first.string() + "InputArrayAdapter", std::move( adapter ) );
+				network->addConnection( ShaderNetwork::Connection(
+					{ adapterHandle, g_arrayOutputNames[ adapterArraySize ] },
+					{ s.first, splineValuesName }
+				) );
+
+				inserted.first->second.first = adapterHandle;
+				inserted.first->second.second = knotOffset;
+			}
+			else
+			{
+				adapterHandle = inserted.first->second.first;
+				knotOffset = inserted.first->second.second;
+			}
+
+			// Default is guaranteed not to be used, since from_chars must either set it or set the error code,
+			// but apparently the compiler doesn't know that
+			int elementId = -1;
+			const char *idStart = &destName[splineElementMatch.position(2)];
+			const char *idEnd = idStart + splineElementMatch.length(2);
+			auto elementIdResult = std::from_chars( idStart, idEnd, elementId );
+
+			if( elementIdResult.ec == std::errc::invalid_argument || elementIdResult.ptr != idEnd )
+			{
+				throw IECore::Exception( "Invalid spline point index " + splineElementMatch[2] );
+			}
```

Comment:
This might just be me, but I really struggle to keep functions this large in my head, and needing 10 lines to do something conceptually simple like convert a string to an int makes it much harder. Personally I doubt performance matters enough here that just `std::stoi( splineElementMatch.str( 2 ) )` wouldn't be perfectly fine. But if we had a tighter regex we could drop all the error checking gumpf here at least.